packages/loopring_v3/contracts/aux/access/LoopringIOExchangeOwner.sol

```diff
@@ -135,7 +155,31 @@ contract LoopringIOExchangeOwner is SelectorBasedAccessManager, ChiDiscount, ERC
         // Process the callback logic.
         _beforeBlockSubmission(blocks, config);
 
+        // Do flash deposits
+        for (uint i = 0; i < flashConfig.deposits.length; i++) {
+            IExchangeV3(target).flashDeposit(
+                flashConfig.deposits[i].to,
+                flashConfig.deposits[i].token,
+                flashConfig.deposits[i].amount
+            );
+        }
+
         target.fastCallAndVerify(gasleft(), 0, decompressed);
+
+        // Do vault logic
+        for (uint i = 0; i < flashConfig.vaults.length; i++) {
+            require(flashConfig.vaults[i].to != target, "EXCHANGE_CANNOT_BE_VAULT");
+            (bool success, ) = flashConfig.vaults[i].to.call(flashConfig.vaults[i].data);
+            require(success, "VAULT_CALL_FAILED");
+        }
+
+        // Make sure flash deposits were repaid
+        for (uint i = 0; i < flashConfig.deposits.length; i++) {
+            require(
+                IExchangeV3(target).getAmountFlashDeposited(flashConfig.deposits[i].token) == 0,
```

Comment:
What if we change 0 to nonzero value? 
```
IExchangeV3(target).getAmountFlashDeposited(flashConfig.deposits[i].token) <= credit[token],
```

This allow us the transaction to success and we just need to manage these credits ...

Reply:
I don't think it's a good idea to leave non-existing tokens in the exchange, I think that's much too risky for users because there is no guarantee it would be paid back and the exchange can be left in an insolvent state. 

Is this to solve the cases where we get less tokens than expected? In general we should do the L1 swap at a price that we are sure will get us a bit more tokens than needed to repay the the flash deposit. Doing that will also build up a bit of surplus we can use in the cases where we need a bit more tokens than we expected. I think this is similar to your `credit` system without passing any of the risk to the users. At the start we could also deposit a small amount of tokens ourselves to build up this reserve.